julia/libCEED.jl/src/Basis.jl

```diff
@@ -0,0 +1,314 @@
+             on the reference element $[-1, 1]$.
+- `qweight`: Array of length `nqpts` holding the quadrature weights on the
+             reference element.
+"""
+function create_h1_basis(
+    c::Ceed,
+    topo::Topology,
+    ncomp,
+    nnodes,
+    nqpts,
+    interp,
+    grad,
+    qref,
+    qweight,
+)
+    dim = getdimension(topo)
+    @assert size(interp) == (nqpts, nnodes)
+    @assert size(grad) == (dim, nqpts, nnodes)
+    @assert length(qref) == nqpts
+    @assert length(qweight) == nqpts
+
+    # Convert from Julia matrices and tensors (column-major) to row-major format
+    interp_rowmajor = collect(interp')
+    grad_rowmajor = permutedims(grad, [3, 2, 1])
+
+    ref = Ref{C.CeedBasis}()
+    C.CeedBasisCreateH1(
+        c[],
+        topo,
+        ncomp,
+        nnodes,
+        nqpts,
+        interp_rowmajor,
+        grad_rowmajor,
+        qref,
+        qweight,
+        ref,
+    )
+    Basis(ref)
+end
+
+"""
+    apply!(b::Basis, nelem, tmode::TransposeMode, emode::EvalMode, u::AbstractCeedVector, v::AbstractCeedVector)
+
+Apply basis evaluation from nodes to quadrature points or vice versa, storing
+the result in the [`CeedVector`](@ref) `v`.
+
+`nelem` specifies the number of elements to apply the basis evaluation to; the
+backend will specify the ordering in CeedElemRestrictionCreateBlocked()
+
+Set `tmode` to `CEED_NOTRANSPOSE` to evaluate from nodes to quadrature or to
+`CEED_TRANSPOSE` to apply the transpose, mapping from quadrature points to
+nodes.
+
+Set the [`EvalMode`](@ref) `emode` to:
+- `CEED_EVAL_NONE` to use values directly,
+- `CEED_EVAL_INTERP` to use interpolated values,
+- `CEED_EVAL_GRAD` to use gradients,
+- `CEED_EVAL_WEIGHT` to use quadrature weights.
+"""
+function apply!(
+    b::Basis,
+    nelem,
+    tmode::TransposeMode,
+    emode::EvalMode,
+    u::AbstractCeedVector,
+    v::AbstractCeedVector,
+)
+    C.CeedBasisApply(b[], nelem, tmode, emode, u[], v[])
+end
+
+"""
+apply(c::Ceed, b::Basis, u::AbstractVector; nelem=1, tmode=NOTRANSPOSE, emode=EVAL_INTERP)
+
+Performs the same function as the above-defined [`apply!`](@ref apply!(b::Basis,
+nelem, tmode::TransposeMode, emode::EvalMode, u::AbstractCeedVector,
+v::AbstractCeedVector)), but automatically convert from Julia arrays to
+[`CeedVector`](@ref) for convenience.
+
+The result will be returned in a newly allocated array of the correct size.
+"""
+function apply(
+    c::Ceed,
+    b::Basis,
+    u::AbstractVector;
+    nelem=1,
+    tmode=NOTRANSPOSE,
+    emode=EVAL_INTERP,
+)
+    u_vec = CeedVector(c, u)
+
+    len_v = (tmode == TRANSPOSE) ? getnumnodes(b) : getnumqpts(b)
+    if emode == EVAL_GRAD
+        len_v *= getdimension(b)
+    end
+
+    v_vec = CeedVector(c, len_v)
+
+    apply!(b, nelem, tmode, emode, u_vec, v_vec)
+    Vector(v_vec)
+end
+
+"""
+    getdimension(b::Basis)
+
+Return the spatial dimension of the given [`Basis`](@ref).
+"""
```

Comment:
I still feel like the tests would be better in the docstrings, like in the Rust interface. It would make the documentation better for the users and make it easier to add tests as we add functionality, in my opinion. I think it would be preferable to do the same for the Python interface as well. (Really, I should do that when I get the chance. Then the Python, Julia, and Rust interfaces would have example code in the auto documentation)

```suggestion
```jldoctest
# create basis
c = Ceed()
dim = 3
b = create_tensor_h1_lagrange_basis(c, dim, 1, 4, 6, GAUSS_LOBATTO)

# verify dimension
@assert getdimension(b) == dim
# output

``` # extra comment to force GitHub code suggestion to show this
"""

Reply:
You can add 
```
DocMeta.setdocmeta!(libCEED, :DocTestSetup, :(using libCEED); recursive = true)
```
to avoid some repeated code in each doctest

Reply:
This is a good point, and having the tests in the docstrings could be helpful. I think they could be useful to supplement the more traditional types of tests. In particular, some of tests have somewhat lengthy setup that would require duplication in the docstrings, and don't really lend themselves so well to that kind of testing.

Reply:
Yeah, I'm not wild about showing the full setup some of the functions need. I think individual lines that don't help much can be hidden:
```
```jldoctest
ceed = Ceed() # hide
# create basis
dim = 3
basis = create_tensor_h1_lagrange_basis(ceed, dim, 1, 2, 3, GAUSS)

# verify dimension
@assert getdimension(basis) == dim

# output

``` # extra comment for markdown
```

the user will then only see
```
# create basis
dim = 3
basis = create_tensor_h1_lagrange_basis(ceed, dim, 1, 2, 3, GAUSS)

# verify dimension
@assert getdimension(basis) == dim

# output

``` # extra comment for markdown
```

Reply:
Related - it seem like the Julia community is a fan of
```
function Base.getproperty(basis::Basis, foo::Symbol)
    if foo == :dim
        return getdimension(basis)
    elseif ....
```
so that
```
@assert basis.dim == 3
```
works.

Thoughts?

Reply:
Yep, we can add those if we want. It hides what is really going on a little bit, which has the potential to be confusing, but I'm fine with it.

Re: `dim` vs. `dimension`, I think this particular abbreviation is fine (see e.g. `Base.ndims`, `Base.permutedims`, etc.)

Reply:
For the doctests, I think it makes sense to identify functions for which the documentation itself will be improved by adding an example that serves as a test (your choice of `getdimension` is one of them). In this case, it's helpful to have a doctest.

However, in the case of a test which is useful for testing the code, but which is not so helpful from an explanatory perspective, I think a traditional test is a better match.